### PR TITLE
Add shift function

### PR DESCRIPTION
**Pull request by tkerola**
This PR implements the shift function, proposed in the following paper: https://arxiv.org/abs/1711.08141

It is a useful alternative to spatial convolutions for reducing the number of model parameters.

### CODE REVIEW

File: chainer/functions/connection/shift.py [hunk 1]
```diff
@@ -0,0 +1,142 @@
...
+
+def _pair(x):
+    if hasattr(x, '__getitem__'):
+        return x
+    return x, x
+
+
+if cuda.available:
+    cupy = cuda.cupy
+    shift_gpu = cupy.ElementwiseKernel(
```
Comment (niboshi) [inline chainer/functions/connection/shift.py]: It slows down `import chainer` if this is run in the toplevel code. How about wrapping with a function?

File: chainer/functions/connection/shift.py [hunk 2]
```diff
@@ -0,0 +1,142 @@
...
+           if (in_row >= 0 && in_row < h && in_col >= 0 && in_col < w) {
+             y = x[b0 * c * h * w + c0 * h * w + in_row * w + in_col];
+           } else {
+             y = 0;
+           }
+        ''',
+        'shift_gpu')
+
+
+class ShiftFunction(function_node.FunctionNode):
```
Comment (niboshi) [inline chainer/functions/connection/shift.py]: Please remove `Function` suffix.

File: chainer/functions/connection/shift.py [hunk 3]
```diff
@@ -0,0 +1,142 @@
...
+
+    def __init__(self, ksize=3, dilate=1):
+        super(ShiftFunction, self).__init__()
+        self.kh, self.kw = _pair(ksize)
+        assert self.kh % 2 == 1, 'kh must be odd'
+        assert self.kw % 2 == 1, 'kw must be odd'
```
Comment (niboshi) [inline chainer/functions/connection/shift.py]: Please raise `ValueError` instead of `assert`.

File: chainer/functions/connection/shift.py [hunk 4]
```diff
@@ -0,0 +1,142 @@
...
+            x_type.shape[1] >= self.kh * self.kw,
+        )
+
+    def forward_cpu(self, inputs):
+        x = inputs[0]
+        b, c, h, w = x.shape
+        py = self.kh // 2 * abs(self.dy)
+        px = self.kw // 2 * abs(self.dx)
+        x = chainer.functions.pad(x, ((0, 0), (0, 0), (py, py), (px, px)),
+                                  'constant')
```
Comment (niboshi) [inline chainer/functions/connection/shift.py]: How about rewriting `forward_cpu` without using Chainer functions?

**Comment by tkerola** [comment]
Thank you! I updated the PR according to your comments.

**Comment by bkvogel** [comment]
Thank you for the PR. Could you also consider adding other float types to the tests: `numpy.float16`, `numpy.float32`, `numpy.float64`.

**Comment by tkerola** [comment]
Sure, I added more tests.

**Comment by bkvogel** [comment]
Sorry for the delay in finishing the review. LGTM!

**Pull request merged**